src/Reflection/RequireExtension/RequireExtendsOrImplementsMethodsClassReflectionExtension.php

```diff
@@ -0,0 +1,81 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Reflection\RequireExtension;
+
+use PHPStan\Analyser\OutOfClassScope;
+use PHPStan\Reflection\ClassReflection;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\MethodsClassReflectionExtension;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\VerbosityLevel;
+
+class RequireExtendsOrImplementsMethodsClassReflectionExtension implements MethodsClassReflectionExtension
+{
+
+	/** @var array<string, array<string, true>> */
+	private array $inProcess = [];
+
+	public function hasMethod(ClassReflection $classReflection, string $methodName): bool
+	{
+		return $this->findMethod($classReflection, $methodName) !== null;
+	}
+
+	public function getMethod(ClassReflection $classReflection, string $methodName): MethodReflection
+	{
+		$method = $this->findMethod($classReflection, $methodName);
+		if ($method === null) {
+			throw new ShouldNotHappenException();
+		}
+
+		return $method;
+	}
+
+	private function findMethod(ClassReflection $classReflection, string $methodName): ?MethodReflection
+	{
+		$extendsOrImplementsTags = array_merge($classReflection->getRequireExtendsTags(), $classReflection->getRequireImplementsTags());
+		foreach ($extendsOrImplementsTags as $extendsOrImplementsTag) {
+			$type = $extendsOrImplementsTag->getType();
+
+			$typeDescription = $type->describe(VerbosityLevel::typeOnly());
+			if (isset($this->inProcess[$typeDescription][$methodName])) {
+				continue;
+			}
+
+			$this->inProcess[$typeDescription][$methodName] = true;
+
+			if (!$type->hasMethod($methodName)->yes()) {
+				unset($this->inProcess[$typeDescription][$methodName]);
+				continue;
+			}
+
+			$method = $type->getMethod($methodName, new OutOfClassScope());
+
+			unset($this->inProcess[$typeDescription][$methodName]);
+
+			// XXX require-implements cannot grant access to static methods?
+			$static = $method->isStatic();
+
+			if (
+				!$static
+				&& $classReflection->hasNativeMethod('__callStatic')
+			) {
+				$static = true;
+			}
+
+			return new RequireExtendsOrImplementsMethodReflection($method, $static);
```

Comment:
You don't need to create new class here, just get the method reflection from the type.